In [56]:
import pandas as pd

Create Dataframes for each csv file

In [13]:
visits = pd.read_csv('visits.csv')
cart = pd.read_csv('cart.csv',
                   parse_dates=[1])
checkout = pd.read_csv('checkout.csv',
                       parse_dates=[1])
purchase = pd.read_csv('purchase.csv',
                       parse_dates=[1])

Print out each dataframe, checking the first five rows with head()

In [107]:
print(visits.head())
print(cart.head())
print(checkout.head())
print(purchase.head())
print(cart.cart_time.count())

                                user_id           visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00
                                user_id           cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00
                                user_id       checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd4

Merge dataframes visits and cart, then print out new merged dataframe

In [108]:
visits_carts = visits.merge(cart, how='left')
print(visits_carts.head())

                                user_id           visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00   

            cart_time  
0                 NaT  
1 2017-01-26 14:44:00  
2 2017-08-20 08:31:00  
3                 NaT  
4                 NaT  


Check the quantity of null values in cart_time to see how many visits did not result in shoppers placing items in their cart

In [98]:
print(visits_carts.cart_time.isnull().sum())

1652


Therefore 1652 rows, of the 2000 rows based on visits df, are null values in cart_time. Which means 348 values in cart_time have a value. This makes sense, cart df is only 348 rows long.

Count total visits

In [99]:
print(visits_carts.user_id.count())

2000


In [100]:
visits_carts[visits_carts.cart_time.isnull()]

user_id           visit_time cart_time
0     943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00       NaT
3     6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00       NaT
4     a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00       NaT
5     7bc62113-d51d-4e6b-85e0-1cf29e3de74a  2017-06-03 20:05:00       NaT
6     6f22cd44-bc63-4449-a9af-2651859da2f9  2017-03-18 23:09:00       NaT
...                                    ...                  ...       ...
1995  33913ac2-03da-45ae-8fc3-fea39df827c6  2017-03-25 03:29:00       NaT
1996  4f850132-b99d-4623-80e6-6e61d003577e  2017-01-08 09:57:00       NaT
1997  f0830b9b-1f5c-4e74-b63d-3f847cc6ce70  2017-09-07 12:56:00       NaT
1998  b01bffa7-63ba-4cd3-9d93-eb1477c23831  2017-07-20 04:37:00       NaT
1999  0336ca81-8d68-443f-9248-ac0b8ad147d5  2017-11-15 10:11:00       NaT

[1652 rows x 3 columns]

Work out the percent of visits that did not result in shoppers placing an item in their cart:

In [101]:
percent_visits_to_carts = (visits_carts.cart_time.isnull().sum()/visits_carts.user_id.count()) * 100

In [137]:
print(str(percent_visits_to_carts) + '% of shoppers after visiting do not place a t-shirt in their cart')

82.6% of shoppers after visiting do not place a t-shirt in their cart


Repeat the left merge for cart and checkout and count null values. What percentage of users put items in their cart, but did not proceed to checkout?

In [114]:
carts_checkout = pd.merge(cart, checkout, how='left').drop_duplicates(subset='user_id')
print(carts_checkout)

                                  user_id           cart_time  \
0    2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
3    4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00   
4    a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00   
5    b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00   
6    a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00   
..                                    ...                 ...   
475  952be3a5-9240-484f-845c-a6b2cbed72ee 2017-05-02 05:40:00   
477  59d2b685-383a-4fce-a325-35d985713a8b 2017-01-20 03:58:00   
478  a1203970-92ae-4fc7-843e-e1740c1be9d5 2017-11-12 23:07:00   
479  55de53ae-89f6-43ff-a26d-df284678aca1 2017-03-17 19:56:00   
480  0ea4cc68-dae4-4e35-b3e0-f0889932e1b5 2017-05-12 08:53:00   

          checkout_time  
0   2017-11-07 21:14:00  
3                   NaT  
4   2017-03-04 11:04:00  
5   2017-09-27 08:26:00  
6                   NaT  
..                  ...  
475 2017-05-02 05:50:00  
477        

What percentage of users put items in their cart, but did not proceed to checkout?

In [143]:
print(carts_checkout.checkout_time.isnull().sum())
print(carts_checkout.user_id.count())

122
348


In [144]:
Percent_carts_checkout = (carts_checkout.checkout_time.isnull().sum()/carts_checkout.user_id.count())*100
print(str(int(Percent_carts_checkout)) + '% of shoppers who put an item in their cart, but did not proceed to checkout')

35% of shoppers who put an item in their cart, but did not proceed to checkout


In [119]:
df_merge = visits.merge(cart, how='left').merge(checkout, how='left').merge(purchase, how='left').drop_duplicates(subset='user_id')
print(df_merge.head())

                                user_id           visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00   

            cart_time       checkout_time       purchase_time  
0                 NaT                 NaT                 NaT  
1 2017-01-26 14:44:00 2017-01-26 14:54:00 2017-01-26 15:08:00  
2 2017-08-20 08:31:00                 NaT                 NaT  
3                 NaT                 NaT                 NaT  
4                 NaT                 NaT                 NaT  



What percentage of users proceeded to checkout, but did not purchase a t-shirt?

In [138]:
print(df_merge.user_id.count() - df_merge['purchase_time'].isnull().sum())
print(df_merge.user_id.count() - df_merge['checkout_time'].isnull().sum())

144
226


In [139]:
percent_checkout_purchase = (df_merge.user_id.count() - df_merge['purchase_time'].isnull().sum())/(df_merge.user_id.count() - df_merge['checkout_time'].isnull().sum())*100

In [142]:
print(str(int(percent_checkout_purchase)) + '% of shoppers who proceeded to checkout and then did not purchase a t-shirt')

63% of shoppers who proceeded to checkout and then did not purchase a t-shirt


82.6% of shoppers after visiting do not place a t-shirt in their cart
35% of shoppers who put an item in their cart, but did not proceed to checkout